File: dpcm/src/main/java/org/sisto/dpcm/DPCM.java
/*
 * DPCM
 * Copyright (C) 2014 Jari Kuusisto
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */
package org.sisto.dpcm;

import org.sisto.dpcm.group.GroupMember;
import org.sisto.dpcm.util.Config;
import org.sisto.dpcm.util.Util;

public class DPCM {
	
	public static final String MAJOR_VERSION 	=	"1"; // major version updates
	public static final String MINOR_VERSION 	= 	"0"; // minor updates
	public static final String MAINT_VERSION 	= 	"0"; // for maintenance updates
	public static final Config ConfigService = Config.getConfig();
	private static final int EXIT_SUCCESS = 0;
	private static final String EMPTY_ARG = "";
	private static GroupMember gm;
	
	public static void main (String[] args) {
		if (printVersion(args)) {
			System.exit(DPCM.EXIT_SUCCESS);
		} else if (printHelp(args)) {
			System.exit(DPCM.EXIT_SUCCESS);
		} else {
			DPCM.shutdownHook();
		}
		init((args.length == 1) ? new String(args[0]) : Util.randomName());
	}
	
	private static void init(String name) {
		ConfigService.loadProperties();
		gm = new GroupMember(name);
		gm.initMember();
		System.out.println(String.format("Running DPCM instance %s...", name));
	}
	
	private static void lize() {
		gm.lizeMember();
	}
	
	private static void shutdownHook() {
		Runtime.getRuntime().addShutdownHook(new Thread() {
		    public void run() {
		    	lize();
		    	System.out.println("\nShutting down! Bye!\n");
		    }
		 });
	}
	
	private static boolean printVersion(String[] args) { 
		boolean exit = false;
		String arg = (args.length == 1) ? new String(args[0]) : EMPTY_ARG;
		
		if (arg.equals("-v") || args.equals("--version")) {
			final String version = String.format("Domain Process Controller Manager version %s", dpcmVersion());
			System.out.println(version);
			exit = true;
		} else {
			exit = false;
		}
		
		return exit;
	}
	
	private static boolean printHelp(String[] args) {
		boolean exit = false;
		String arg = (args.length == 1) ? new String(args[0]) : EMPTY_ARG;
		
		if (arg.equals("-h") || args.equals("--help")) {
			final String help = String.format("Usage: java -jar dpcm.jar [<group-member-name>]");
			System.out.println(help);
			exit = true;
		} else {
			exit = false;
		}
		
		return exit;
	}
	
	private static String dpcmVersion() {
		final String version = String.format("%s.%s.%s", DPCM.MAJOR_VERSION, DPCM.MINOR_VERSION, DPCM.MAINT_VERSION);
		
		return version;
	}
}

File: dpcm-wildfly/src/main/resources/dpcm.js
/*
 * DPCM
 * Copyright (C) 2014 Jari Kuusisto
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */
importPackage(org.jboss.as.cli.scriptsupport)      
importPackage(java.lang)
/*
 * Note! You can also run CLI-scripts as a batch by typing in console:
 *       'run-batch --file <filename.cli>'
 */
cli = null

function connect() {
	ct = false
	cli = CLI.newInstance()
	
	try {
		cli.connect()
		ct = true
	} catch (e) {
		println("scripting error, could not connect")
		print(e)
		ct = false
	}
	
	return ct
}

function disconnect() {
	dt = false
	
	try {
		cli.disconnect()
		dt = true
	} catch (e) {
		println("scripting error, could not disconnect")
		dt = false
	}
	
	return dt
}

function isDomainController() {
	ret = false
	connected = connect()
	
	if (connected) {
		result = cli.cmd(":read-attribute(name=process-type)")
		response = result.getResponse()
		dc = response.get("result").asString()
		if (dc == "Domain Controller") {
			ret = true
		} else {
			ret = false
		}
	} else {
		ret = false
	}
	disconnect()
	
	return ret
}

function isProcessRunning() {
	ret = false
	connected = connect()
	
	if (connected) {
		result = cli.cmd(":read-attribute(name=name)")
		ret = result.isSuccess()
	} else {
		ret = false
	}
	disconnect()
	
	return ret
}

function commitSuicideMan() {
	ret = false
	connected = connect()
	
	if (connected) {
		result = cli.cmd(":read-attribute(name=local-host-name)")
		response = result.getResponse()
		localhost = response.get("result").asString()
		result = cli.cmd("/host="+localhost+":shutdown(restart=false)")
		shuttingdown = result.isSuccess()
		if (shuttingdown) {
			ret = true
		} else {
			ret = false
		}
	} else {
		ret = false
	}
	disconnect()
	
	return ret
}

